### ISSUE TITLE
[PPO强化学习做市] | PPO Market Making

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4501]
**标签**: quantitative-strategy
### 摘要
基于Proximal Policy Optimization (PPO)强化学习算法，构建具有竞争力的电子合约做市策略。

### 算法描述
**核心原理：**
PPO是一种策略梯度强化学习算法，通过限制策略更新幅度实现稳定训练。在做市场景中，Agent学习动态调整买卖价差和报价数量，以最大化收益同时控制风险。

**核心概念：**
- State: 订单簿深度、市场波动率、持仓 Greeks、库存不平衡度
- Action: 买入价偏移、卖出价偏移、报价数量
- Reward: 价差收益 - 库存成本 - 滑点损失

**数学模型：**
```
策略网络：π(a|s) = softmax(W2·tanh(W1·s + b1) + b2)

价值网络：V(s) = w2·tanh(w1·s + b1)

PPO Loss:
L_clip = E[min(r_t(θ)·A_t, clip(r_t(θ), 1-ε, 1+ε)·A_t)]

其中r_t(θ) = π(a|s) / π_old(a|s)，使用clipped surrogate objective
```

**库存风险惩罚：**
```
InventoryPenalty = λ × |Q - Q_target| × σ_market

其中Q为当前持仓，σ_market为市场波动率
```

### Python代码示例
```python
import numpy as np
import torch
import torch.nn as nn
from typing import Dict, List, Tuple
from dataclasses import dataclass

@dataclass
class MarketState:
    mid_price: float
    bid_depth: np.ndarray  # 买1-买5深度
    ask_depth: np.ndarray  # 卖1-卖5深度
    inventory: float
    volatility: float
    spread: float
    time_to_expiry: float

class PPOMarketMaker:
    def __init__(
        self,
        state_dim: int = 15,
        action_dim: int = 4,
        hidden_dim: int = 128,
        lr_actor: float = 3e-4,
        lr_critic: float = 1e-3,
        gamma: float = 0.99,
        epsilon: float = 0.2,
        lam: float = 0.95
    ):
        self.state_dim = state_dim
        self.action_dim = action_dim
        self.gamma = gamma
        self.epsilon = epsilon
        self.lam = lam
        
        # Actor-Critic网络
        self.actor = self._build_actor(hidden_dim)
        self.critic = self._build_critic(hidden_dim)
        self.actor_optimizer = torch.optim.Adam(self.actor.parameters(), lr=lr_actor)
        self.critic_optimizer = torch.optim.Adam(self.critic.parameters(), lr=lr_critic)
        
        self.buffer = []
        self.training_history = []
        
    def _build_actor(self, hidden_dim: int) -> nn.Module:
        return nn.Sequential(
            nn.Linear(self.state_dim, hidden_dim),
            nn.Tanh(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.Tanh(),
            nn.Linear(hidden_dim, self.action_dim),
            nn.Softmax(dim=-1)
        )
    
    def _build_critic(self, hidden_dim: int) -> nn.Module:
        return nn.Sequential(
            nn.Linear(self.state_dim, hidden_dim),
            nn.Tanh(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.Tanh(),
            nn.Linear(hidden_dim, 1)
        )
    
    def get_state_features(self, market_state: MarketState) -> np.ndarray:
        """特征工程：提取状态特征"""
        features = []
        
        # 归一化价格
        features.append(market_state.mid_price / 100.0)
        
        # 订单簿深度特征
        bid_imbalance = (market_state.bid_depth.sum() - market_state.ask_depth.sum()) / \
                        (market_state.bid_depth.sum() + market_state.ask_depth.sum() + 1e-8)
        features.append(bid_imbalance)
        
        # 各档位深度比率
        for i in range(min(5, len(market_state.bid_depth))):
            depth_ratio = market_state.bid_depth[i] / (market_state.ask_depth[i] + 1e-8)
            features.append(depth_ratio)
        
        # 库存状态
        features.append(market_state.inventory / 100.0)
        
        # 波动率
        features.append(market_state.volatility / 0.3)
        
        # 价差
        features.append(market_state.spread / 0.01)
        
        # 时间衰减
        features.append(market_state.time_to_expiry / 86400.0)
        
        return np.array(features, dtype=np.float32)
    
    def select_action(self, state: np.ndarray, deterministic: bool = False) -> Tuple[int, float]:
        """选择动作"""
        state_tensor = torch.FloatTensor(state).unsqueeze(0)
        
        with torch.no_grad():
            probs = self.actor(state_tensor).numpy()[0]
        
        if deterministic:
            action = np.argmax(probs)
            log_prob = np.log(probs[action] + 1e-8)
        else:
            action = np.random.choice(len(probs), p=probs)
            log_prob = np.log(probs[action] + 1e-8)
        
        return action, log_prob
    
    def calculate_reward(
        self,
        spread: float,
        inventory: float,
        volatility: float,
        executed: bool,
        trade_direction: int
    ) -> float:
        """计算奖励"""
        # 价差收益
        spread_profit = spread * (1 if executed else 0)
        
        # 库存成本
        inventory_cost = -0.5 * volatility * abs(inventory)
        
        # 方向惩罚（避免过度做空或做多）
        directional_penalty = -0.01 * abs(trade_direction) if abs(inventory) > 50 else 0
        
        return spread_profit + inventory_cost + directional_penalty
    
    def compute_gae(
        self,
        rewards: List[float],
        values: List[float],
        dones: List[bool]
    ) -> Tuple[List[float], List[float]]:
        """计算GAE（广义优势估计）"""
        advantages = []
        last_gae = 0
        
        for t in reversed(range(len(rewards))):
            if t == len(rewards) - 1:
                next_value = 0
            else:
                next_value = values[t + 1]
            
            delta = rewards[t] + self.gamma * next_value * (1 - dones[t]) - values[t]
            last_gae = delta + self.gamma * self.lam * (1 - dones[t]) * last_gae
            advantages.insert(0, last_gae)
        
        returns = [adv + val for adv, val in zip(advantages, values)]
        return advantages, returns
    
    def update_policy(
        self,
        states: List[np.ndarray],
        actions: List[int],
        rewards: List[float],
        dones: List[bool]
    ):
        """PPO策略更新"""
        if len(states) < 32:
            return
        
        # 转为tensor
        states_t = torch.FloatTensor(np.array(states))
        actions_t = torch.LongTensor(actions)
        
        # 计算values
        with torch.no_grad():
            values = [self.critic(torch.FloatTensor(s).unsqueeze(0)).item() for s in states]
        
        # 计算GAE
        advantages, returns = self.compute_gae(rewards, values, dones)
        advantages_t = torch.FloatTensor(advantages)
        returns_t = torch.FloatTensor(returns)
        
        # 归一化优势
        advantages_t = (advantages_t - advantages_t.mean()) / (advantages_t.std() + 1e-8)
        
        # PPO更新
        for _ in range(10):
            # 获取当前策略概率
            probs = self.actor(states_t)
            dist = torch.distributions.Categorical(probs)
            log_probs = dist.log_prob(actions_t)
            entropy = dist.entropy().mean()
            
            # 重要性采样比率
            ratio = torch.exp(log_probs - torch.FloatTensor([self.buffer[i][1] for i in range(len(states))]))
            
            # Clipped surrogate objective
            surr1 = ratio * advantages_t
            surr2 = torch.clamp(ratio, 1 - self.epsilon, 1 + self.epsilon) * advantages_t
            actor_loss = -torch.min(surr1, surr2).mean() - 0.01 * entropy
            
            # Critic loss
            values_pred = self.critic(states_t).squeeze()
            critic_loss = nn.MSELoss()(values_pred, returns_t)
            
            # 更新
            self.actor_optimizer.zero_grad()
            self.critic_optimizer.zero_grad()
            actor_loss.backward()
            critic_loss.backward()
            self.actor_optimizer.step()
            self.critic_optimizer.step()
        
        self.buffer.clear()


if __name__ == \"__main__\":
    agent = PPOMarketMaker(state_dim=15, action_dim=4)
    
    # 模拟市场状态
    market_state = MarketState(
        mid_price=100.0,
        bid_depth=np.array([10, 8, 6, 4, 2]),
        ask_depth=np.array([9, 7, 5, 3, 1]),
        inventory=20.0,
        volatility=0.02,
        spread=0.01,
        time_to_expiry=3600
    )
    
    state = agent.get_state_features(market_state)
    print(f\"State shape: {state.shape}\")
    
    action, log_prob = agent.select_action(state)
    print(f\"Action: {action}, Log Prob: {log_prob:.4f}\")


### 数据来源
- Reddit: r/quant, r/algorithmictrading
- 小红书: 量化投资
- 知乎: 量化交易
